### PR TITLE
auth0: implement directory provider

### PR DESCRIPTION
**Pull request by grounded042**
## Summary
This PR adds the Auth0 directory provider implementation. I have not updated the docs but can either do that in this PR, or in a new PR.

## Related issues
Based on #1453 

**Checklist**:
- [X] add related issues
- [ ] updated docs
- [X] updated unit tests
- [ ] updated UPGRADING.md
- [X] ready for review

### CODE REVIEW

**Comment by grounded042** [comment]
@travisgroth sorry about that, I should have noticed. It has been removed!

**Review by calebdoxsey — COMMENTED** [review]
Some minor things and a couple questions, but otherwise this LGTM. Thanks for submitting.

File: internal/directory/auth0/auth0.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+				Id:   *role.ID,
+				Name: *role.Name,
+			})
+		}
+
+		page++
+		shouldContinue = listRes.HasNext()
+	}
+
+	return roles, nil
```
[the remark this answers is not in the document]
Comment (calebdoxsey) [reply internal/directory/auth0/auth0.go:151]: (I don't think them being unsorted actually hurts anything, but it helps with consistency for testing, etc)

[the remark this answers is not in the document]
Comment (grounded042) [reply internal/directory/auth0/auth0.go:151]: I will check into this.

File: internal/directory/auth0/auth0.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+		for _, user := range usersRes.Users {
+			ids = append(ids, *user.ID)
+		}
+
+		page++
+		shouldContinue = usersRes.HasNext()
+	}
+
+	return ids, nil
```
[the remark this answers is not in the document]
Comment (grounded042) [reply internal/directory/auth0/auth0.go:174]: I will check into this.

File: internal/directory/auth0/auth0.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	"fmt"
+	"sort"
+
+	"github.com/rs/zerolog"
+
+	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/pkg/grpc/databroker"
+	"github.com/pomerium/pomerium/pkg/grpc/directory"
+
+	"gopkg.in/auth0.v4/management"
```
[the remark this answers is not in the document]
Comment (grounded042) [reply internal/directory/auth0/auth0.go]: done!

File: internal/directory/auth0/auth0.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+		}
+
+		page++
+		shouldContinue = listRes.HasNext()
+	}
+
+	return roles, nil
+}
+
+func getRoleUserIDs(rm RoleManager, id string) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (grounded042) [reply internal/directory/auth0/auth0.go]: done!

**Review by calebdoxsey — APPROVED** [review]

File: internal/directory/auth0/auth0.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (grounded042) [reply internal/directory/auth0/auth0.go:150]: The roles are not sorted by ID from the API.

File: internal/directory/auth0/auth0.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (grounded042) [reply internal/directory/auth0/auth0.go:173]: They are not sorted here either.

**Comment by calebdoxsey** [comment]
Thanks @grounded042 . I went ahead and merged the changes.

**Pull request merged**